### ISSUE TITLE
understanding ensembling

### ISSUE THREAD

**Issue by Howuhh** [patrick-kidger/equinox#359]
I am a relative newcomer to jax and many things are still a mystery to me. Right now I'm trying to write a simple RL algorithm that uses ensembles, so I need to be able to train ensembles. After reading the documentation and trying a couple of examples, I'm left with a lot of questions about how vmap works in this case.  I'd welcome hints or links where this is explained in more detail.

For example, I want to train an ensemble of networks on batches of data. Creation is straightforward, we just vectorise over keys. 
```python
# from https://docs.kidger.site/equinox/tricks/
key = jax.random.PRNGKey(0)
keys = jax.random.split(key, 8)

@eqx.filter_vmap
def make_ensemble(key):
    return eqx.nn.MLP(2, 2, 2, 2, key=key)

mlp_ensemble = make_ensemble(keys)
```

However, forward pass vectorization over batches is confusing. This will work for batches:
```python
@eqx.filter_vmap(in_axes=dict(x=None))
def evaluate_ensemble(model, x):
    return jax.vmap(model)(x)
```
Conceptually, I understand that this function must be vectorized by the number of models, and x will be broadcasted. Why then there is a dict? Why such approaches does not work? Doesn't we need to specify all arguments as it is done in jax documentation?
```python
@eqx.filter_vmap(in_axes=(0, None))
# or
@eqx.filter_vmap(in_axes=(0, None), out_axes=(0, None))
```
Also, can we vectorise it over batch without inner vmap?

P.S. there is a typo in the documentation, seem like `eqx.if_array` should be `eqx.is_array`

**Comment by patrick-kidger** [comment]
Using a dictionary for `in_axes` declares that the named axes should have the specified batch axes -- and that anything not declared should have the default. The default being `0` if it's an array, and `None` if it's not an array. (This is a syntax specific to `equinox.filter_vmap`, it's not part of `jax.vmap`.)

Moreover, it really is an `if_array`, not `is_array`. These are different functions! The former is the default behaviour discussed above: it declares a `0` batch axis if something is an array, and `None` otherwise.

Thus
```python
@eqx.filter_vmap(in_axes=dict(x=None))
def evaluate_ensemble(model, x):
    return jax.vmap(model)(x)
```
is equivalent to
```python
@eqx.filter_vmap(in_axes=dict(model=eqx.if_array(0), x=None))  # explicitly fill in the default for `model`
def evaluate_ensemble(model, x):
    return jax.vmap(model)(x)
```

Note that your snippet
```python
@eqx.filter_vmap(in_axes=dict(x=None))
def evaluate_ensemble(model, x):
    return jax.vmap(model)(x)
```
does something different again: as you have two vmaps then the two are composed, and you will end up with the combination of every model being evaluated against every x.

**Comment by Howuhh** [comment]
Yes, that makes a lot more sense, thank you!
